Overhaul of query handling — `q` permalinks, `session[:query_record]`

nimmolo opened the pull request:
### `session[:query_record]`
This PR totally changes how MO keeps track of a user's current query, which is how they can step through results, see associated records, map them, etc. The big change is storing the `QueryRecord` `id` in the session, rather than manually passing the alphabetized `id` through the url as a `q` param, as we currently do. The `QueryRecord` is created when the query is first executed (e.g., by an `:index`), and is thereafter available without being manually passed.

The issue with the current approach is that the `QueryRecord` is perishable (like a session), but its ID is encoded in the URL like a permalink. It currently has to be passed through the url so that a `show` page, for example, can know what the previous/next records were, or what filtered index to go back to. 

But the obvious place to store a reference that needs to be ubiquitously available, like the current user, is the `session` (i.e. a cookie).

### `q` permalinks
For links or requests that need to modify the query (a search or some other filtered link), they can now send a `q` param in the URL that will function as a shareable permalink.

On the coding side, with this change we can now build a query simply by writing a params hash for `q` into a link. The params and values obviously have to be params for the relevant `Query` class, but these are all documented. The following will produce a permalink and the desired results. 
```ruby
link_to(
  observations_path(
    params: { 
      q: { model: :Observation, 
           by_users: [1],  locations: [720], 
           names: { lookup: "Hygrocybe felix", include_synonyms: true }, 
           projects: [330], order_by: "confidence" }
    } 
  )
)
```
No need to resort to “pattern search” permalinks.

### Details:
- MO will no longer have a `q` param for most urls. Only on indexes, maps, downloads, etc. - pages that are the results of queries. The `q` param in those cases will be a Rails-encoded hash that describes the query, and is used by the receiving action to generate the results. The hash is barely legible, but the important thing is that it's parseable by Rails without us doing anything. If you call `params[:q]` in the receiving action, you will get the exact hash you sent, with the keys already symbolized.
- When a user requests an index page with a `q` param, either as a result of a search or a permalink, the index action saves a description of the query in a `QueryRecord`, as currently. The change is that the record ID is stored in `session[:query_record]` so it doesn’t have to be passed in the URL. Any action in the app has access to the session, so they have access to the same `QueryRecord` previously passed in the `q` param. (Why didn’t we do this before?) Tons and tons of repetitive code has been deleted.
- No more `pass_query_params`, anywhere. Far fewer calls to `add_query_param`, `redirect_with_query`, `link_with_query`, `icon_link_with_query` etc.
- For backward compatibility, indexes can still handle incoming "alphabetized" `q` params. Incoming `q=34iuy` params will get immediately updated to permalinks before being passed to the next page, so they will get phased out pretty quickly.  Tested.
- Adds a temporary `permalink` column to QueryRecord that we can write if the record's URL is now a permalink. QueryRecords are updated every time the alphabetized `q` hits a page - that's how we know if they're still in use or whether to delete them. When `QueryRecord.where(permalink: false).count == 0` and is stable for awhile, we can delete the code that handles alphabetized `q`.
- Why not just store the `q` hash in the session directly, rather than a reference to the QueryRecord? Because the encoded hash might exceed the allowable size of a cookie.
- I also want to keep the `QueryRecord` mechanism intact, because that's how we could store multiple queries per user, e.g. a Names query, an Observations query, or several of each. Outside the present scope

nimmolo committed:
First pass

nimmolo committed:
fix current_query_is_rss_log

nimmolo committed:
Fix some tests

nimmolo committed:
Update names_helper.rb

nimmolo committed:
more tests

nimmolo committed:
Update search_controller_test.rb

nimmolo committed:
Update queries.rb

nimmolo committed:
Update images_controller_test.rb

nimmolo committed:
Eliminate use of `@controller.query_params` in tests. Use `get_query_param`

nimmolo committed:
Fix add_query_param and append_query_param_to_path

nimmolo committed:
Refactor show/prev links

nimmolo committed:
Update show_prev_next_helper.rb

nimmolo committed:
Update queries.rb

nimmolo committed:
Fix more tests

nimmolo committed:
debugging

ImagesControllerTest#test_destroy_image_with_query demonstrates what's wrong with encoding. `q` is getting double-encoded as a query string when generated from the `@query_params` variable.

nimmolo committed:
Get rid of `query_params` method, which seemed to double-encode q

Also convert `pass_query_params` ivar @query_params to @query_param (from hash to string)

nimmolo committed:
Try not encoding

nimmolo committed:
Update sorter_helper.rb

nimmolo committed:
Update queries.rb

nimmolo committed:
Sort out q swap via Stimulus, add `qs-esm`

nimmolo committed:
Fix query_from_q_param

nimmolo committed:
Improve param permitting

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

nimmolo committed:
Abandon permitting params, just use `to_unsafe_hash`

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

nimmolo committed:
Fix downloads controller test

nimmolo committed:
Update downloads_controller_test.rb

nimmolo committed:
Update collection_numbers_controller_test.rb

nimmolo committed:
Update maps_controller_test.rb

nimmolo committed:
Update collection_numbers_controller_test.rb

nimmolo committed:
Update herbarium_records_controller_test.rb

nimmolo committed:
Add query_string method to all controller tests

nimmolo committed:
Update downloads_controller_test.rb

nimmolo committed:
Update herbarium_records_controller_test.rb

nimmolo committed:
Update collection_numbers_controller_test.rb

nimmolo committed:
Update sequences_controller_test.rb

nimmolo committed:
Update maps_controller_test.rb

nimmolo committed:
Update remove_observations_controller_test.rb

nimmolo committed:
Update remove_observations_controller_test.rb

nimmolo committed:
Update search_controller_test.rb

nimmolo committed:
Update observations_controller_index_test.rb

nimmolo committed:
Fix set_project_ivar

nimmolo committed:
Update controller_extensions.rb

Fix url_for override

nimmolo committed:
Update images_controller_test.rb

nimmolo committed:
Refactor names_controller_test, fix

nimmolo committed:
Update _form.erb

nimmolo committed:
Update articles_controller_test.rb

nimmolo committed:
Fix results of find_or_create_query

nimmolo committed:
Clean up

Don't forget to store "permalink" on new QR's

nimmolo committed:
Merge branch 'nimmo-break-up-names-controller-test' into nimmo-q-permalinks

nimmolo committed:
Add test for QueryRecord.check_param

nimmolo committed:
Add column to QR to keep track of updated permalinks

nimmolo committed:
Test for permalinks and refactor queryrecord model methods

nimmolo committed:
Cover multiple-observations warning in CollectionNumbersController

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

nimmolo committed:
add hr form multiple test too

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

mo-nathan commented on app/controllers/application_controller/queries.rb:
```diff
@@ -335,9 +384,9 @@ def redirect_to_next_object(method, model, id)
-                                  id: id }, query))
+    redirect_to({ controller: object.show_controller,
+                  action: object.show_action,
+                  id:, q: get_query_param(query) })
```
Is there a plan to rename this method?  I'm not a fan of method names that start with "get_" and I believe avoiding them is also standard Rails practice.  Is there a compelling reason it's not just called `q_param`?  It could even be off the Query class which might be a better place than ApplicationController::Queries.  Doesn't need to be this PR, but seems like a reasonable cleanup to do at some point.

mo-nathan commented on app/controllers/application_controller/queries.rb:
```diff
@@ -195,23 +252,24 @@ def query_from_session
-    end
-  end
-  # helper_method :query_params
+  # def query_params(query = nil)
```
Delete?

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

nimmolo committed:
first changes, mostly in ap ctrlr/queries

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

nimmolo committed:
Merge branch 'nimmo-q-permalinks' into nimmo-q-permalinks-session-query-record

nimmolo committed:
Don't store_query_in_session on query_params_set

nimmolo committed:
Update show_prev_next_helper.rb

nimmolo committed:
Update query-results_controller.js

Fix var name

nimmolo committed:
Update test methods `parse_query_params`

nimmolo committed:
Merge branch 'nimmo-q-permalinks' into nimmo-q-permalinks-session-query-record

nimmolo committed:
query_params_set restore store_query_in_session

nimmolo committed:
Update namings_integration_test.rb

nimmolo committed:
Update indexes.rb

Comments

nimmolo committed:
Update queries.rb

pass query params back from add_user_content_filter_parameters
query_needs_update?
query_from_q_param refactor

nimmolo committed:
Merge branch 'nimmo-q-permalinks' into nimmo-q-permalinks-session-query-record

nimmolo committed:
Update lurker_integration_test.rb

Split up test_obs_at_location_index

nimmolo committed:
current_query

Check for @query first

nimmolo committed:
Update show_prev_next_helper.rb

show_page_incoming_query(object) - check for session[:query_record]

nimmolo committed:
Update capybara_session_extensions.rb

comment

nimmolo committed:
Update query-results_controller.js

Fix q setting

nimmolo committed:
Update query-results_controller.js

Fix q setting

nimmolo committed:
Update sorter_helper.rb

Make sorter send proper query links
Handle them in order_by_or_flash_if_unknown (coming under :q not :by)

nimmolo committed:
Update matrix_box_helper.rb

Remove link_with_query!

nimmolo committed:
Update sequence.rb

New methods first_id, prev_id, next_id, last_id

nimmolo committed:
Update show_prev_next_helper.rb

Use query methods to avoid query params in links

nimmolo committed:
Update query_test.rb

test new Query methods

nimmolo committed:
Update collection_numbers_helper.rb

Remove add_query_param

nimmolo committed:
Sketch out converting type filters to query params

Need to handle like :by

nimmolo committed:
Got it

nimmolo committed:
Update comments_helper.rb

nimmolo committed:
Update descriptions_helper.rb

nimmolo committed:
Update general_helper.rb

nimmolo committed:
Update herbaria_helper.rb

nimmolo committed:
Update herbarium_records_helper.rb

nimmolo committed:
Update images_helper.rb

nimmolo committed:
Update rss_logs_controller.rb

remove debugger

nimmolo committed:
Update locations_helper.rb

nimmolo committed:
Update names_helper.rb

nimmolo committed:
Update namings_helper.rb

nimmolo committed:
Update observations_helper.rb

nimmolo committed:
Update projects_helper.rb

nimmolo committed:
Update species_lists_helper.rb

nimmolo committed:
Update descriptions.rb

nimmolo committed:
Update interests_controller.rb

nimmolo committed:
Update lookups_controller.rb

nimmolo committed:
Update destroy.rb

nimmolo committed:
Update descriptions_helper.rb

nimmolo committed:
Update lightbox_helper.rb

nimmolo committed:
Update link_helper.rb

nimmolo committed:
Update names_helper.rb

nimmolo committed:
Update namings_helper.rb

nimmolo committed:
Update projects_helper.rb

nimmolo committed:
Update descriptions_helper.rb

nimmolo committed:
Update _form.erb

nimmolo committed:
Update _form.erb

nimmolo committed:
Update _form.erb

nimmolo committed:
Update _vote_panel.html.erb

nimmolo committed:
Update _form.erb

nimmolo committed:
Update _form.html.erb

nimmolo committed:
Update new.html.erb

nimmolo committed:
Update _form.html.erb

nimmolo committed:
Update _form.erb

nimmolo committed:
Update _form.erb

nimmolo committed:
Update _observation.erb

nimmolo committed:
Update _form.erb

nimmolo committed:
Descriptions object_path & _with_query, use show_link_args

nimmolo committed:
Update projects_controller.rb

nimmolo committed:
Remove a whole mess of query params from 111 files

nimmolo committed:
Update glossary_terms_controller_test.rb

nimmolo committed:
Update projects_controller_test.rb

nimmolo committed:
Update _object.html.erb

nimmolo committed:
Update observations_helper.rb

nimmolo committed:
Fix some controller tests

nimmolo committed:
Adjust more controller tests

nimmolo committed:
Update indexes.rb

nimmolo committed:
Fix hidden href on disabled prev/next arrow links

nimmolo committed:
Update lurker_integration_test.rb

nimmolo committed:
Update destroy.rb

nimmolo committed:
Update observations_integration_test.rb

nimmolo committed:
Remove duplicate `sorted_index_opts`

nimmolo committed:
Add tests for new q: { type: } } links

nimmolo committed:
Fix default and deep_merge sorted_index_opts

nimmolo committed:
Update contributors_controller.rb

deep_merge sorted_index_opts

nimmolo committed:
Update rss_logs_controller.rb

deep_merge sorted_index_opts

nimmolo committed:
Adjust for new default order (:date)

nimmolo committed:
No more `pass_query_params`

nimmolo committed:
Delete query-results_controller.js

nimmolo committed:
Merge pull request #3238 from MushroomObserver/nimmo-q-permalinks-session-query-record

Switch to use `session[:query_record]`. Turn off long permalinks where not needed to reproduce query results

nimmolo committed:
Switch to `add_q_param` and `q_param` (previously `get_query_param`)

nimmolo committed:
Whoops - replace `full_q_param` with `q_param`

nimmolo committed:
Finish replacing `full_q_param`

nimmolo committed:
Update queries.rb

nimmolo committed:
Merge pull request #3240 from MushroomObserver/nimmo-q-permalinks-session-query-record

Switch from `add_query_param` to `add_q_param` etc

nimmolo committed:
Merge branch 'nimmo-q-permalinks' of https://github.com/MushroomObserver/mushroom-observer into nimmo-q-permalinks

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks

nimmolo committed:
`query_params_set` to `update_stored_query`

nimmolo committed:
Update names_controller_index_test.rb

JoeCohen commented on app/classes/query/modules/sequence.rb:
```diff
@@ -131,6 +131,12 @@ def first
     new_self
   end
 
+  def first_id
```
1. Is this dead code? The only call I could find was QueryTest, line 428
2. If it's not dead, will this work:

```suggestion
    result_ids.&first
```

JoeCohen commented on app/classes/query/modules/sequence.rb:
```diff
@@ -131,6 +131,12 @@ def first
   end
 
+  def first_id
+    return nil unless result_ids.length
+
+    result_ids.first
```
Is this method dead code? The only call I could find was QueryTest, line 428
If it's not dead, will this work:

```suggestion
    result_ids.&first
```

nimmolo replied:
Thanks. It may be dead code, but it's new code not old.

I wrote new `prev_id` and `next_id` methods and am using them in the `show_prev_next` UI and elsewhere, and i figured I'd need a `first_id` and `last_id` method to tell if we're at the first or last records. But I'm able to get it another way. 

I'll commit this suggestion but may end up deleting both.

nimmolo committed:
Update sequence.rb

nimmolo committed:
Update queries.rb

mo-nathan commented:
Crashing bug:

1. From the home page click Search.
2. Search for “Amanita”.
3. Click on an observation.
4. Click on the Observations Index icon.
5. Enter “10" in the pages.
6. Click “Go” icon.
7. Boom!

```
TypeError in ObservationsController#index
no implicit conversion of Symbol into Integer
Extracted source (around line #240):

  def query_from_q_param_hash
    q_param = params[:q]
    return nil if q_param[:model].blank? # <=== Failing line

    Query.lookup(q_param[:model].to_sym,
                 **q_param.except(:model).to_unsafe_hash)

Rails.root: /Users/nathan/src/mushroom-observer

Application Trace | Framework Trace | Full Trace
app/controllers/application_controller/queries.rb:240:in `[]'
app/controllers/application_controller/queries.rb:240:in `query_from_q_param_hash'
app/controllers/application_controller/queries.rb:215:in `query_from_q_param'
app/controllers/application_controller/indexes.rb:176:in `order_by_or_flash_if_unknown'
app/controllers/application_controller/indexes.rb:168:in `sorted_index_opts'
app/controllers/application_controller/indexes.rb:154:in `sorted_index'
app/controllers/application_controller/indexes.rb:53:in `block in build_index_with_query'
app/controllers/application_controller/indexes.rb:49:in `each'
app/controllers/application_controller/indexes.rb:49:in `build_index_with_query'
app/controllers/observations_controller/index.rb:7:in `index'
app/controllers/application_controller.rb:148:in `catch_errors_and_log_request_stats'
```

nimmolo committed:
Update index_pagination_helper.rb

nimmolo committed:
goto_page_input - use q_param_to_hidden_fields

nimmolo committed:
Update index_pagination_helper.rb

Do not pass the :id through to next/prev page.

nimmolo committed:
Update index_pagination_helper.rb

Letters form should use the same

nimmolo committed:
Update filters_helper.rb

Use Set in case there are doubles, as for many Genus names

nimmolo committed:
Add tests for urls in pager/pagination

nimmolo committed:
Update observations_controller_index_test.rb

Test for concise filters

nimmolo committed:
Tests of :q

nimmolo committed:
Query method q_param

nimmolo commented:
@mo-nathan After several false starts, i actually have this fixed.

I added tests checking the url format and making sure the `:id` param is not there on next/prev page links.

Also added `q_param` as a method of a `Query` instance.

nimmolo committed:
Merge branch 'main' into nimmo-q-permalinks
